profiling/service/service.go

```diff
@@ -0,0 +1,78 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package service
+
+import (
+	"context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/grpclog"
+
+	"google.golang.org/grpc/internal/profiling"
+	ppb "google.golang.org/grpc/profiling/proto"
+	pspb "google.golang.org/grpc/profiling/proto/service"
+)
+
+type ProfilingConfig struct {
+	Enabled         bool
+	StreamStatsSize uint32
+	Server          *grpc.Server
+}
+
+func registerService(s *grpc.Server) {
+	grpclog.Infof("registering profiling service")
+	pspb.RegisterProfilingServer(s, &profilingServer{})
+}
+
+func Init(pc *ProfilingConfig) (err error) {
+	if err = profiling.InitStats(pc.StreamStatsSize); err != nil {
+		return
+	}
+
+	registerService(pc.Server)
+
+	// Do this last after everything has been initialised and allocated.
+	profiling.SetEnabled(pc.Enabled)
+
+	return
+}
+
+type profilingServer struct{}
+
+func (s *profilingServer) SetEnabled(ctx context.Context, req *pspb.SetEnabledRequest) (ser *pspb.SetEnabledResponse, err error) {
+	grpclog.Infof("processing SetEnabled (%v)", req.Enabled)
+	profiling.SetEnabled(req.Enabled)
+
+	ser = &pspb.SetEnabledResponse{Success: true}
+	err = nil
+	return
+}
+
+func (s *profilingServer) GetStreamStats(req *pspb.GetStreamStatsRequest, stream pspb.Profiling_GetStreamStatsServer) (err error) {
+	grpclog.Infof("processing stream request for stream stats")
+	results := profiling.StreamStats.Drain()
+	grpclog.Infof("stream stats size: %v records", len(results))
+
+	for i := 0; i < len(results); i++ {
```

Comment:
Does it make sense to define a `defaultBatchSize` and send that many stat protos in one response? If not, that's totally fine too.

Reply:
Do you mean something like pagination?

Reply:
I didn't have anything fancy in mind. Just wanted to see if we can avoid a `Send` call for each `StatProto` and batch them up. But then, you would need to change the service definition to have a repeated field. Might not be worth the effort I guess.

Reply:
An array of batched stats looks more natural, I agree. One downside is that it may result in response sizes in the 100MB+ range for a few hundred thousand RPCs. Streams feel more "correct" in this case?

Reply:
Streams are definitely better.